aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,125 @@
 
-|stub-icon| Model Validation
-============================
+In this article:
 
-.. include:: /../common/stub-notice.txt
+- `Introduction to model validation`_
+- `Data annotations`_
+- `Model state`_
+- `Custom validation`_
+- `Client side validation`_ 
+- `Remote valiation`_
 
-.. _issue: https://github.com/aspnet/Docs/issues/116
+Introduction to model validation
+--------------------------------
+Before an app stores data in a database, the app must validate the data. Data must be scrubbed for security, verified that it is appropriately formatted by type and size, and it must conform to your rules. Validation is necessary although it can be redundant and tedious to implement. In MVC, validation happens on both the client and server. 
+
+By using data annotations in MVC apps you can validate data by customizing models so they mimic the underlying structure of the data store. This way you can enforce integrity rules and constraints against models by applying attributes to model properties. These attributes generate validation code, thereby reducing the amount of code you must write. Fortunately, .NET has abstracted validation into the data annotations, often reducing validation to a single line of code.
+
+Data annotations
+----------------
+Data annotations are a way to configure model validation so it's similar to validation on fields in tables. This includes constraints such as assigning data types, required fields, or a primary or foreign key. Other types of validation include applying patterns to data to enforce integrity, such as a credit card, phone number, or email address (regular expressions). Data annotations do all this for you automatically.
+
+Below is an annotated ``Movie`` model from an app that stores information about movies and TV shows. Most of the fields are required, the ``Id`` field is a primary key, and several string properties have length requirements. Additionally, there is a numeric range in place for the ``Price`` property from 0 to $999.99, along with two custom annotations.
+
+.. literalinclude:: validation/sample/Movie.cs
+   :language: c#
+   :lines: 7-37
+   :dedent: 4
+
+Simply reading through the model reveals the rules about data for this app, making it easier to maintain the code. Below are several popular built-in data annotations:
+
+- ``[Bind]``: Determines whether the property is included or excluded from model binding.
+- ``[CreditCard]``: Formats the property as a credit card type.
+- ``[Compare]``: Compares two properties in a model. It works with various types, such as int, string, or date. 
+- ``[DataType]``: Applies extra type information to a property. For example, a string property may be formatted as an email data type.
+- ``[DisplayName]``: Determines the text that renders in form field labels and validation error messages.
+- ``[Email]``: Formats the property with an email format.
+- ``[ForeignKey]``: Provides foreign key functionality.
+- ``[Key]``: Provides primary key functionality. 
```

Comment:
MVC ignores both `[ForeignKey]` and `[Key]`. Likely relevant for EF but not here.